### PR TITLE
CI Multiarch

### LINKED ISSUE #3112: tpm2_load: fails due to 64 bit openssl call
TPM2_LOAD fails while retrieving parent value from a TSS Privkey PEM file. Code calls to ASN1_INTEGER_GET_UNIT64 from the openssl API, but the function seems to not work correctly on a 32 bit system.

`
persistent-handle: 0x81100000
action: persisted
ERROR:esys:src/tss2-esys/esys_mu.c:813:iesys_MU_IESYS_RESOURCE_Unmarshal() Error unmarshaling subfield name ErrorCode (0x00090006) 
ERROR:esys:src/tss2-esys/esys_tr.c:96:Esys_TR_Deserialize() Unmarshal resource object ErrorCode (0x00090006) 
ERROR: Esys_TR_Deserialize(0x90006) - mu:A buffer isn't large enough
ERROR: Incorrect handle value, got: "mykey.pem", expected expected [o|p|e|n|l] or a handle number
ERROR: Unable to convert parent to integer/value too large
ERROR: Cannot make sense of object context "mykey.pem"
ERROR: Unable to run load
tpm2 load -r $pem_file.pem -c $pem_file.ctx on line 90 failed: 1
`

### LINKED ISSUE #3118: tpm2_encodeobject creates negative value for parent handle in 32 bit system
tssprivkey formatted pem generation on 32 bit system creates negative parent handle value for persistent objects. This happens when persistent object address space of 81xx_xxxx is parsed as a signed integer and the MSB state determines the value to be negative.

### PR DESCRIPTION
**Pull request by malikabhi05**
Fixes #3118
Fixes #3112

### CODE REVIEW

**Comment by idesai** [comment]
@JuergenReppSIT can you please comment on the failures in fapi tests seen [here](https://github.com/tpm2-software/tpm2-tools/runs/7796693036?check_suite_focus=true#step:5:238).

File: test/integration/helpers.sh [hunk 1]
```diff
@@ -292,6 +292,7 @@ function start_sim() {
         case "$TPM2_SIM" in
             *swtpm) "$TPM2_SIM" socket --tpm2 --server port="$tpm2_sim_port" \
                                        --ctrl type=tcp,port="$tpm2_sim_cmd_port" \
+                                       --seccomp action="none" \
```
Comment (malikabhi05) [inline test/integration/helpers.sh:295]: @idesai

**Comment by idesai** [comment]
> @malikabhi05 @idesai Could you please run the tss2_provisioning of one fapi test e.g. fapi-get-random.sh with:
> 
> ```
> ls -l */* >&2
> TSS2_LOG=fapi+trace tss2_provision
> ```
> 
> The profile, which was not found, should be written to the tmp directory of the test and listed by the ls command. The part of the log to find the profile should be similar to:
> 
> ```
> trace:fapi:src/tss2-fapi/ifapi_io.c:489:ifapi_io_dirfiles() List directory: /tmp/tpm2_test_ux0guC/tss2_fapi.KWILpL/
> ...
> trace:fapi:src/tss2-fapi/ifapi_io.c:502:ifapi_io_dirfiles() Looking at P_ECCP256SHA256.json 
> ```

This is on the failing build

trace:fapi:src/tss2-fapi/ifapi_io.c:489:ifapi_io_dirfiles() List directory: /tmp/tpm2_test_W77OAn/tss2_fapi.JEKMv5/ 
ERROR:fapi:src/tss2-fapi/ifapi_profiles.c:116:ifapi_profiles_initialize_async() No files found in profile dir /tmp/tpm2_test_W77OAn/tss2_fapi.JEKMv5/ that match the pattern P_*.json 
ERROR:fapi:src/tss2-fapi/api/Fapi_Initialize.c:283:Fapi_Initialize_Finish() Read profile ErrorCode (0x0006000b) 
trace:fapi:src/tss2-fapi/api/Fapi_Initialize.c:85:Fapi_Initialize() finished 
Fapi_Initialize(0x6000B) - fapi:A parameter has a bad value
TSS2_LOG=fapi+trace tss2_provision on line 21 failed: 1
ERROR:fapi:src/tss2-fapi/ifapi_profiles.c:116:ifapi_profiles_initialize_async() No files found in profile dir /tmp/tpm2_test_W77OAn/tss2_fapi.JEKMv5/ that match the pattern P_*.json 
ERROR:fapi:src/tss2-fapi/api/Fapi_Initialize.c:283:Fapi_Initialize_Finish() Read profile ErrorCode (0x0006000b) 
Fapi_Initialize(0x6000B) - fapi:A parameter has a bad value
FAIL test/integration/fapi/fapi-get-random.sh (exit status: 1)

**Comment by malikabhi05** [comment]
@JuergenReppSIT we're seeing this on a 32 bit container "ubuntu-20.04.arm32v7"

File: test/integration/helpers.sh (hunk 1 above)
Comment (malikabhi05) [reply test/integration/helpers.sh:295]: @williamcroberts @idesai should I only disable seccomp for multi arch builds or for all builds?

Comment (williamcroberts) [reply test/integration/helpers.sh:295]: All builds is fine

**Comment by malikabhi05** [comment]
load test will pass once #3112 is updated and merged.

File: Makefile.am [hunk 2]
```diff
@@ -315,6 +316,10 @@ test_unit_test_tpm2_eventlog_LDADD = $(CMOCKA_LIBS) $(LDADD)
 test_unit_test_tpm2_eventlog_yaml_CFLAGS = $(AM_CFLAGS) $(CMOCKA_CFLAGS)
 test_unit_test_tpm2_eventlog_yaml_LDADD = $(CMOCKA_LIBS) $(LDADD)
 
+test_unit_test_object_CFLAGS = $(AM_CFLAGS) $(CMOCKA_CFLAGS)
+test_unit_test_object_LDFLAGS = -Wl,--wrap=tpm2_util_object_fetch_tssprivkey_from_file
```
Comment (idesai) [inline Makefile.am:320]:
```suggestion
```

File: test/unit/test_object.c [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+#include <setjmp.h>
+#include <cmocka.h>
+#include <tss2/tss2_mu.h>
+#include "object.h"
+
+
+//tool_rc tpm2_util_object_fetch_tssprivkey_from_file(const char *objectstr, BIO **input_bio, TSSPRIVKEY_OBJ **tpk);
+//tool_rc __real_tpm2_util_object_fetch_tssprivkey_from_file(const char *objectstr, BIO **input_bio, TSSPRIVKEY_OBJ **tpk);
+
+tool_rc __wrap_tpm2_util_object_fetch_tssprivkey_from_file(const char *objectstr,
```
Comment (idesai) [inline test/unit/test_object.c:16]:
```suggestion
tool_rc tpm2_util_object_fetch_tssprivkey_from_file(const char *objectstr,
```

File: test/unit/test_object.c [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        printf("QQQQQQQQQQQQQQ!!\n");
+        printf("----------------------\n");
+        /*
+         * Create a blob and copy over to tpk such that tpk->parent is negative
+         */
+        TSSPRIVKEY_OBJ *tpk_internal = TSSPRIVKEY_OBJ_new();
+        tpk_internal->parent->flags = 0x0;
+        tpk_internal->parent->type = 0x102; /* 9th bit is for sign maybe */
+        tpk_internal->parent->length = 0x4;
+        unsigned char address[] = {0x7e, 0xe0, 0x00, 0x01};
```
Comment (idesai) [inline test/unit/test_object.c:31]: Use ASN1_Integer_set routines like so in **tpm2_encodeobject**

File: test/unit/test_object.c [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+int main(int argc, char* argv[]) {
+    (void) argc;
+    (void) argv;
+
+    const struct CMUnitTest tests[] = {
+        cmocka_unit_test(test_tpm2_util_object_fetch_parent_from_tpk)
+    };
+
+    return cmocka_run_group_tests(tests, NULL, NULL);
+}
```
Comment (idesai) [inline test/unit/test_object.c:69]:
```suggestion
}

```

File: lib/object.c [hunk 6]
```diff
@@ -173,32 +174,37 @@ static tool_rc tpm2_util_object_setup_primary(ESYS_CONTEXT *esys_ctx,
...
+tool_rc tpm2_util_object_fetch_parent_from_tpk(const char *objectstr,
     uint64_t *val) {
 
     tool_rc rc = tool_rc_general_error;
     BIO *input_bio = 0;
     TSSPRIVKEY_OBJ *tpk = 0;
     BIGNUM *bn_parent = 0;
-    tpm2_util_object_fetch_tssprivkey_from_file(objectstr, &input_bio, &tpk);
-    if (!input_bio || !tpk) {
+    tool_rc rct = tpm2_util_object_fetch_tssprivkey_from_file(objectstr, &input_bio, &tpk);
```
Comment (idesai) [inline lib/object.c:206]:
```suggestion
rc = tpm2_util_object_fetch_tssprivkey_from_file(objectstr, &input_bio, &tpk);

File: test/unit/test_object.c [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+#include <setjmp.h>
+#include <cmocka.h>
+#include <tss2/tss2_mu.h>
+#include "object.h"
+
+
+//tool_rc tpm2_util_object_fetch_tssprivkey_from_file(const char *objectstr, BIO **input_bio, TSSPRIVKEY_OBJ **tpk);
+//tool_rc __real_tpm2_util_object_fetch_tssprivkey_from_file(const char *objectstr, BIO **input_bio, TSSPRIVKEY_OBJ **tpk);
+
```
Comment (idesai) [inline test/unit/test_object.c:15]:
```suggestion
```

Comment (idesai) [inline test/unit/test_object.c:15]:
```suggestion
```

File: test/integration/tests/encodeobject.sh [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+tpm2 createprimary -G $alg_pem_load -C o \
+-a "fixedtpm|fixedparent|sensitivedataorigin|userwithauth|restricted|decrypt|noda" \
+-c $pem_parent.ctx
+
+tpm2 create -C $pem_parent.ctx -u $pem_file.pub -r $pem_file.priv
+
+tpm2 evictcontrol -C o -c $pem_parent.ctx 0x81100000
+
+tpm2 encodeobject -C 0x81100000 -u $pem_file.pub -r $pem_file.priv \
+-o $pem_file.pem
```
Comment (idesai) [inline test/integration/tests/encodeobject.sh:44]: $Pem_persistent_parent instead of 0x81100000?

File: test/integration/tests/encodeobject.sh [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+tpm2 load -r $pem_file.pem -c $pem_file.ctx
+
+PARENTVAL=`openssl asn1parse -in $pem_file.pem -inform pem | awk '{print $7}'`
+if egrep -q "^:-[0-9a-fA-F]{8}" <<< "$PARENTVAL"
+then
+  # negative parent exists
```
Comment (idesai) [inline test/integration/tests/encodeobject.sh:51]:
```suggestion
  # negative parent exists
  echo "Fail: Parent value is negative"
```

File: test/integration/tests/encodeobject.sh [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+  exit 1
+else
+  # negative parent does not exist
+  exit 0
```
Comment (idesai) [inline test/integration/tests/encodeobject.sh:55]:
```suggestion
```

File: test/integration/tests/encodeobject.sh [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+fi
+
```
Comment (idesai) [inline test/integration/tests/encodeobject.sh:57]:
```suggestion

exit 0

```

**Review by idesai — CHANGES_REQUESTED** [review]

File: .github/workflows/main.yml [hunk 12]
```diff
@@ -43,6 +43,35 @@ jobs:
...
+      - name: Setup QEMU
+        run: |
+          sudo apt-get update
+          sudo apt-get install qemu binfmt-support qemu-user-static
+          docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
+      - name: Check out repository
+        uses: actions/checkout@v2
+      - name: Launch Action
+        uses:
+          malikabhi05/test-ac/runCI@master
```
Comment (idesai) [inline .github/workflows/main.yml:66]:
```suggestion
          tpm2-software/ci/runCI@main
```

**Review by idesai — CHANGES_REQUESTED** [review]

**Comment by idesai** [comment]
One small change:
[ci: point CI to correct CI repo](https://github.com/tpm2-software/tpm2-tools/pull/3110/commits/060c4283e2b3f3b9bc60a11cb8277d4b4a156044) must be merged into [Adding multiarch var](https://github.com/tpm2-software/tpm2-tools/pull/3110/commits/6a1c53204a9b769baaf564894b5d10df1acba80c)
The fix in the last commit is for a bug in the same PR.

**Pull request merged**